[WPE] Pack inspector resources in a .gresource file instead of a shared library

siwei-li opened the pull request:
#### 8ceb1da47e75a488ae4c12017a861636904acd4f
<pre>
[WPE] Pack inspector resources in a .gresource file instead of a shared library
<a href="https://bugs.webkit.org/show_bug.cgi?id=186594">https://bugs.webkit.org/show_bug.cgi?id=186594</a>

Reviewed by Adrian Perez de Castro.

Generating .gresource file for WebInspector resources instead of a shared library
for WPE platform; the file is installed to the data directory instead of LIBDIR.

* Source/JavaScriptCore/PlatformWPE.cmake:
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
(Inspector::backendCommands):
* Source/WebKit/InspectorGResources.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/WebProcess/Inspector/wpe/WebInspectorUIWPE.cpp:
(WebKit::WebInspectorUI::didEstablishConnection):

Canonical link: <a href="https://commits.webkit.org/284269@main">https://commits.webkit.org/284269@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78e122634bd81e19757234dd3404983a973d3234

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68790 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48184 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72861 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19936 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19755 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54838 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13261 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71856 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59398 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35295 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40685 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16826 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-grid/subgrid/alignment-in-subgridded-axes-001.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18298 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61907 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62635 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17171 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74556 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68037 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16421 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62314 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12803 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59479 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62350 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10309 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3923 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89816 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10519 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43982 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15910 "Found 42 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Number/property_and_index_of_number.js.default, ChakraCore.yaml/ChakraCore/test/Strings/property_and_index_of_string.js.default, ChakraCore.yaml/ChakraCore/test/es6/regex-quantifiers.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck3.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck6.js.default, ChakraCore.yaml/ChakraCore/test/strict/19.function.js.default, jsc-layout-tests.yaml/js/script-tests/dfg-call-function-hit-watchpoint.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-convert-this-object-then-exit-on-other.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-not-string.js.layout ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45059 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44798 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc reviewed (changes requested):
The changes in the CMake build files are fine, but note that the `libWPEWebInspectorResources.so` library is loaded from `Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp` and `Source/WebKit/WebProcess/Inspector/wpe/WebInspectorUIWPE.cpp`. You will need to update those files (I think they are the only ones) to load the `.gresource` bundle instead using [g_resource_load()](https://docs.gtk.org/gio/type_func.Resource.load.html) + [g_resources_register()](https://docs.gtk.org/gio/func.resources_register.html).

We may want to check an environment variable, for example `WEBKIT_INSPECTOR_RESOURCES`, to check for a non-default path to the `.gresource` file, too.

aperezdc commented on Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
```diff
@@ -42,22 +42,21 @@ GRefPtr<GBytes> backendCommands()
     std::call_once(flag, [] {
-        const char* libDir = PKGLIBDIR;
+        const char* dataDir = PKGDATADIR;
+        GError* error = nullptr;
```
You could use a `GUniqueOutPtr` smart pointer here, and then the `g_clear_error()` call below is not needed.

```suggestion
        GUniqueOutPtr<GError> error;
```

aperezdc commented on Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
```diff
@@ -42,22 +42,21 @@ GRefPtr<GBytes> backendCommands()
-        if (!resourcesModule) {
-            WTFLogAlways("Error loading libWPEWebInspectorResources.so: %s", g_module_error());
+        GUniquePtr<char> gResourceFilename(g_build_filename(dataDir, "InspectorGResourceBundle.gresource", nullptr));
+        GResource* gresource = g_resource_load(gResourceFilename.get(), &error);
```
Please use a `GRefPtr` smart pointer here, to avoid leaking the object:

```suggestion
        GRefPtr<GResource> gresource = adoptGRef(g_resource_load(gResourceFilename.get(), &error));
```

aperezdc commented on Source/WebKit/PlatformWPE.cmake:
```diff
@@ -528,24 +529,9 @@ if (ENABLE_BREAKPAD)
-list(APPEND WPEWebInspectorResources_SYSTEM_INCLUDE_DIRECTORIES
-    ${GLIB_INCLUDE_DIRS}
-)
+WEBKIT_BUILD_INSPECTOR_GRESOURCES(${WebInspectorUI_DERIVED_SOURCES_DIR} "InspectorGResourceBundle.gresource")
```
I think the name of the file could be just `inspector.gresource`, the longer name does not add any additional meaning 😉

aperezdc reviewed (changes requested):
This looks almost ready to go. The logic is right, but let's use smart pointers where possible, I left some comments about this.

mcatanzaro commented on Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
```diff
@@ -42,22 +42,21 @@ GRefPtr<GBytes> backendCommands()
+        if (!gresource) {
+            WTFLogAlways("Error loading InspectorGResourceBundle.gresource: %s", error->message);
+            g_clear_error(&error);
             return;
```
Just use g_error() here, to crash. Surely this should be fatal.

aperezdc commented on Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
```diff
@@ -42,22 +42,19 @@ GRefPtr<GBytes> backendCommands()
-            WTFLogAlways("Error loading libWPEWebInspectorResources.so: %s", g_module_error());
-            return;
+        GUniquePtr<char> gResourceFilename(g_build_filename(dataDir, "inspector.gresource", nullptr));
+        GRefPtr<GResource> gresource = g_resource_load(gResourceFilename.get(), &error.outPtr());
```
This is missing an `adoptGRef()` call:

```suggestion
        GRefPtr<GResource> gresource = adoptGRef(g_resource_load(gResourceFilename.get(), &error.outPtr()));
```

aperezdc commented on Source/WebKit/WebProcess/Inspector/wpe/WebInspectorUIWPE.cpp:
```diff
@@ -36,22 +37,20 @@ void WebInspectorUI::didEstablishConnection()
-            WTFLogAlways("Error loading libWPEWebInspectorResources.so: %s", g_module_error());
-            return;
+        GUniquePtr<char> gResourceFilename(g_build_filename(dataDir, "inspector.gresource", nullptr));
+        GRefPtr<GResource> gresource = g_resource_load(gResourceFilename.get(), &error.outPtr());
```
Here as well, we need to adopt the pointer on construction:

```suggestion
        GRefPtr<GResource> gresource = adoptGRef(g_resource_load(gResourceFilename.get(), &error.outPtr()));
```

aperezdc reviewed (commented):
Looks good, we are only missing adopting the created `GResource` instances with `adoptGRef` and this should be good to go. Please take a look at the comments 🙏🏼

siwei-li committed:
[WPE] Pack inspector resources in a .gresource file instead of a shared library
https://bugs.webkit.org/show_bug.cgi?id=186594

Reviewed by Adrian Perez de Castro.

Generating .gresource file for WebInspector resources instead of a shared library
for WPE platform; the file is installed to the data directory instead of LIBDIR.

* Source/JavaScriptCore/PlatformWPE.cmake:
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
(Inspector::backendCommands):
* Source/WebKit/InspectorGResources.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/WebProcess/Inspector/wpe/WebInspectorUIWPE.cpp:
(WebKit::WebInspectorUI::didEstablishConnection):

Canonical link: https://commits.webkit.org/284269@main

philn commented on Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
```diff
@@ -42,22 +42,19 @@ GRefPtr<GBytes> backendCommands()
-        // Probably no need for a specific env var here. Assume the inspector resources.so file is
-        // in the same directory as the injected bundle lib, for developer builds.
-        const char* path = g_getenv("WEBKIT_INJECTED_BUNDLE_PATH");
+        const char* path = g_getenv("WEBKIT_INSPECTOR_RESOURCES_PATH");
```
afaics none of our tooling is setting this env var. Can you follow-up about this?

siwei-li replied:
No I don't think the env var is set anywhere, the idea was just to let people pass in the variable if the developer mode is enabled, should we add it to scripts for WPE API?

philn replied:
Yes, in a similar manner as `WEBKIT_INJECTED_BUNDLE_PATH` (grep in Tools :))

siwei-li replied:
Just checking in with you guys after this merge: https://github.com/WebKit/WebKit/pull/35403
There's a bunch few other places where `WEBKIT_INJECTED_BUNDLE_PATH` is set, like in the minibrowser and the tests, should I also add `WEBKIT_INSPECTOR_RESOURCES_PATH` in a similar way?
(also @carlosgcampos)

aperezdc replied:
@siwei-li Yes, please go ahead and add the environment variable in the places where it's missing.
